### PR TITLE
feat(terragrunt): add support for tfr prefix

### PR DESCRIPTION
**Pull request by ps-occrp**
## Changes

Adds support for [tfr source](https://terragrunt.gruntwork.io/docs/reference/config-blocks-and-attributes/#terraform)

## Context

Terragrunt manager does not support [tfr source](https://terragrunt.gruntwork.io/docs/reference/config-blocks-and-attributes/#terraform) which has existed in Terragrunt for a few years, this PR adds support for that.

Related to https://github.com/renovatebot/renovate/discussions/22890

## Documentation (please check one with an [x])

- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or

### CODE REVIEW

File: lib/modules/manager/terragrunt/modules.ts [hunk 1]
```diff
@@ -62,7 +62,16 @@ export function analyseTerragruntModule(
     const moduleParts = source.split('//')[0].split('/');
     if (moduleParts[0] === '..') {
       dep.skipReason = 'local';
-    } else if (moduleParts.length >= 3) {
+    } else if (source.startsWith('tfr:///')) {
+      const hostnameMatch = hostnameMatchRegex.exec('registry.terraform.io');
+      if (hostnameMatch?.groups) {
+        dep.registryUrls = [`https://${hostnameMatch.groups.hostname}`];
+      }
```
Comment (rarkins) [inline lib/modules/manager/terragrunt/modules.ts:69]: This result will always be the same, so can be refactored/simplified

**Review by rarkins — CHANGES_REQUESTED** [review]
Please add or update a test

File: lib/modules/manager/terragrunt/modules.ts (hunk 1 above)
Comment (ps-occrp) [reply lib/modules/manager/terragrunt/modules.ts]: Can you point me to place where I can see test for some of the other part of the code? tried to find it myself didn't find it.

**Comment by ps-occrp** [comment]
> Please add or update a test

Can you point me to place where I can see test for some of the other part of the code? tried to find it myself didn't find it.

**Comment by rarkins** [comment]
image

File: lib/modules/manager/terragrunt/modules.ts [hunk 2]
```diff
@@ -62,6 +62,11 @@ export function analyseTerragruntModule(
     const moduleParts = source.split('//')[0].split('/');
     if (moduleParts[0] === '..') {
       dep.skipReason = 'local';
+    } else if (source.startsWith('tfr:///')) {
+      dep.registryUrls = ['registry.terraform.io'];
```
Comment (viceice) [inline lib/modules/manager/terragrunt/modules.ts:66]:
```suggestion
      dep.registryUrls = ['https://registry.terraform.io'];
```
shouldn't it have a https prefix?

isn't this the default registry url of terraform module datasource? so completely remove?

File: lib/modules/manager/terragrunt/extract.spec.ts [hunk 3]
```diff
@@ -7,6 +7,32 @@ describe('modules/manager/terragrunt/extract', () => {
...
+          {
+            datasource: 'terraform-module',
+            depName: 'terraform-aws-modules/vpc/aws',
+            depType: 'terragrunt',
+          },
+          {},
+        ],
+      });
+      expect(res?.deps).toHaveLength(4);
+      expect(res?.deps.filter((dep) => dep.skipReason)).toHaveLength(0);
```
Comment (viceice) [inline lib/modules/manager/terragrunt/extract.spec.ts:33]:
```suggestion
```
the `toEqual` validated this already

**Comment by viceice** [comment]
please create a reproduction and run this against it. we need to validate your slitting is working as expected.

**Comment by ps-occrp** [comment]
> please create a reproduction and run this against it. we need to validate your slitting is working as expected.

Thanks for suggesting this, in reproduction repo I found some missing things, I ended up refactoring it using a regex.

**Comment by ps-occrp** [comment]
@rarkins @viceice can you merge this PR?

**Review by rarkins — APPROVED** [review]

**Pull request merged**